odoo/addons/base/models/ir_attachment.py

```diff
@@ -194,27 +239,17 @@ def _compute_datas(self):
     def _inverse_datas(self):
```

Comment:
The splitting of `_inverse_datas` into several methods should not change the logic and actually only serves for better extensibility.

This solves the following scenario:

The filestore must be checked each time and the file may have to be deleted.

```python
if fname:
   self._file_delete(fname)
```

This is not a major problem yet, because an extension could do the same, but if this extension also has a rule like this and there is another extension, it would have to take it into account or know about one another, but in my eyes this is very hard to maintain.

Before:
```python
def _inverse_datas(self):
        location = self._storage()
        if location == 'lobject':
            for attach in self:
                value = attach.datas
                bin_data = base64.b64decode(value) if value else b''
                vals = {
                    'file_size': len(bin_data),
                    'checksum': self._compute_checksum(bin_data),
                    'index_content': self._index(bin_data, attach.datas_fname, attach.mimetype),
                    'store_fname': False,
                    'db_datas': False,
                    'store_lobject': bin_data,
                }
                fname = attach.store_fname
                super(LObjectIrAttachment, attach.sudo()).write(vals)
                if fname:
                    self._file_delete(fname)
        else:
            super(LObjectIrAttachment, attach)._inverse_datas()
```

After:
```python
api.multi
    def _inverse_datas(self):
        location = self._storage()
        if location == 'lobject':
            for attach in self:
                value = attach.datas
                bin_data = base64.b64decode(value) if value else b''
                vals = self._get_datas_inital_vals()
                vals = self._update_datas_vals(vals, attach, bin_data)
                vals['store_lobject'] = bin_data
                clean_vals = self._get_datas_clean_vals(attach)
                models.Model.write(attach.sudo(), vals)
                self._clean_datas_after_write(clean_vals)
        else:
            super(LObjectIrAttachment, self)._inverse_datas()
```